File: src/main/java/xyz/phanta/cbtweaker/integration/mekanism/buffer/MekanismLaserBuffer.java
package xyz.phanta.cbtweaker.integration.mekanism.buffer;

import com.google.gson.JsonObject;
import io.github.phantamanta44.libnine.event.ModDependentEventBusSubscriber;
import io.github.phantamanta44.libnine.util.data.ByteUtils;
import io.github.phantamanta44.libnine.util.data.ISerializable;
import io.github.phantamanta44.libnine.util.function.ICapabilityInstanceConsumer;
import io.github.phantamanta44.libnine.util.render.DrawOrientation;
import io.github.phantamanta44.libnine.util.render.TextureRegion;
import io.github.phantamanta44.libnine.util.world.BlockSide;
import mekanism.api.lasers.ILaserReceptor;
import mekanism.common.capabilities.Capabilities;
import mezz.jei.api.IJeiHelpers;
import net.minecraft.client.resources.I18n;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import xyz.phanta.cbtweaker.CbtLang;
import xyz.phanta.cbtweaker.CbtMod;
import xyz.phanta.cbtweaker.buffer.BufferFactory;
import xyz.phanta.cbtweaker.buffer.BufferObserver;
import xyz.phanta.cbtweaker.buffer.BufferType;
import xyz.phanta.cbtweaker.buffer.ingredient.IngredientMatcher;
import xyz.phanta.cbtweaker.buffer.ingredient.IngredientMatcherType;
import xyz.phanta.cbtweaker.event.CbtIngredientHandlerRegistrationEvent;
import xyz.phanta.cbtweaker.event.CbtRegistrationEvent;
import xyz.phanta.cbtweaker.gui.CbtTextureResources;
import xyz.phanta.cbtweaker.gui.ScreenRegion;
import xyz.phanta.cbtweaker.gui.component.DoubleEnergyBarGuiComponent;
import xyz.phanta.cbtweaker.gui.inventory.UiElement;
import xyz.phanta.cbtweaker.integration.jei.ingredient.JeiIngredient;
import xyz.phanta.cbtweaker.integration.jei.render.JeiUiElement;
import xyz.phanta.cbtweaker.integration.mekanism.MekanismIntegration;
import xyz.phanta.cbtweaker.integration.mekanism.jei.JeiMekanismJoulesIngredient;
import xyz.phanta.cbtweaker.util.ConfigException;
import xyz.phanta.cbtweaker.util.Positioning;
import xyz.phanta.cbtweaker.util.block.MachineSideHandler;
import xyz.phanta.cbtweaker.util.helper.DataLoadUtils;

import javax.annotation.Nullable;
import java.util.*;
import java.util.function.Supplier;

@ModDependentEventBusSubscriber(dependencies = MekanismIntegration.MOD_ID)
public class MekanismLaserBuffer implements ILaserReceptor, ISerializable {

    public static final ResourceLocation ID = CbtMod.INSTANCE.newResourceLocation("mekanism_laser");

    private final double capacity;
    private double energyStored = 0D;
    @Nullable
    private final BufferObserver observer;

    private final Positioning uiPosition;
    private final TextureRegion bgTexture, fgTexture;
    private final int barOffsetX, barOffsetY;
    private final DrawOrientation barOrientation;

    public MekanismLaserBuffer(double capacity,
                               Positioning uiPosition, TextureRegion bgTexture, TextureRegion fgTexture,
                               int barOffsetX, int barOffsetY, DrawOrientation barOrientation,
                               @Nullable BufferObserver observer) {
        this.capacity = capacity;
        this.observer = observer;
        this.uiPosition = uiPosition;
        this.bgTexture = bgTexture;
        this.fgTexture = fgTexture;
        this.barOffsetX = barOffsetX;
        this.barOffsetY = barOffsetY;
        this.barOrientation = barOrientation;
    }

    public double getCapacity() {
        return capacity;
    }

    public double getEnergyStored() {
        return energyStored;
    }

    public void setEnergyStored(double energyStored) {
        if (this.energyStored == energyStored) {
            return;
        }
        this.energyStored = energyStored;
        if (observer != null) {
            observer.onIngredientsChanged();
        }
    }

    public double insert(double amount, boolean commit) {
        double toTransfer = Math.min(amount, capacity - energyStored);
        if (toTransfer <= 0) {
            return 0;
        }
        if (commit) {
            setEnergyStored(energyStored + toTransfer);
        }
        return toTransfer;
    }

    public double extract(double amount, boolean commit) {
        if (amount <= 0 || energyStored <= 0) {
            return 0;
        }
        double toTransfer = Math.min(amount, energyStored); // must be positive
        if (commit) {
            setEnergyStored(energyStored - toTransfer);
        }
        return toTransfer;
    }

    public Positioning getUiPosition() {
        return uiPosition;
    }

    public TextureRegion getBackgroundTexture() {
        return bgTexture;
    }

    public TextureRegion getForegroundTexture() {
        return fgTexture;
    }

    public int getBarOffsetX() {
        return barOffsetX;
    }

    public int getBarOffsetY() {
        return barOffsetY;
    }

    public DrawOrientation getBarOrientation() {
        return barOrientation;
    }

    @Override
    public void receiveLaserEnergy(double energy, EnumFacing face) {
        setEnergyStored(Math.min(energyStored + energy, capacity));
    }

    @Override
    public boolean canLasersDig() {
        return false;
    }

    public MekanismLaserBuffer copy(@Nullable BufferObserver observer) {
        MekanismLaserBuffer buf = new MekanismLaserBuffer(capacity,
                uiPosition, bgTexture, fgTexture, barOffsetX, barOffsetY, barOrientation, observer);
        buf.energyStored = energyStored;
        return buf;
    }

    @Override
    public void serBytes(ByteUtils.Writer data) {
        data.writeDouble(energyStored);
    }

    @Override
    public void deserBytes(ByteUtils.Reader data) {
        energyStored = data.readDouble();
    }

    @Override
    public void serNBT(NBTTagCompound tag) {
        tag.setDouble("EnergyStored", energyStored);
    }

    @Override
    public void deserNBT(NBTTagCompound tag) {
        energyStored = tag.getDouble("EnergyStored");
    }

    public static class Accumulator {

        private final List<MekanismLaserBuffer> buffers = new ArrayList<>();

        private void accumulate(MekanismLaserBuffer buffer) {
            buffers.add(buffer);
        }

        public double getStored() {
            return buffers.stream().mapToDouble(MekanismLaserBuffer::getEnergyStored).sum();
        }

        public double getCapacity() {
            return buffers.stream().mapToDouble(MekanismLaserBuffer::getCapacity).sum();
        }

        public double insert(double amount, boolean commit) {
            double remaining = amount;
            for (MekanismLaserBuffer buffer : buffers) {
                remaining -= buffer.insert(remaining, commit);
                if (remaining <= 0) {
                    return amount;
                }
            }
            return amount - remaining;
        }

        public double extract(double amount, boolean commit) {
            double remaining = amount;
            for (MekanismLaserBuffer buffer : buffers) {
                remaining -= buffer.extract(remaining, commit);
                if (remaining <= 0) {
                    return amount;
                }
            }
            return amount - remaining;
        }

        public Accumulator copy() {
            Accumulator acc = new Accumulator();
            for (MekanismLaserBuffer buffer : buffers) {
                acc.accumulate(buffer.copy(null));
            }
            return acc;
        }

    }

    public static class JeiBuffer {

        private final double capacity;
        @Nullable
        private JeiMekanismJoulesIngredient contents = null;
        private final Positioning uiPosition;
        private final TextureRegion bgTexture, fgTexture;
        private final int barOffsetX, barOffsetY;
        private final DrawOrientation barOrientation;

        public JeiBuffer(double capacity,
                         Positioning uiPosition, TextureRegion bgTexture, TextureRegion fgTexture,
                         int barOffsetX, int barOffsetY, DrawOrientation barOrientation) {
            this.capacity = capacity;
            this.uiPosition = uiPosition;
            this.bgTexture = bgTexture;
            this.fgTexture = fgTexture;
            this.barOffsetX = barOffsetX;
            this.barOffsetY = barOffsetY;
            this.barOrientation = barOrientation;
        }

        public boolean isEmpty() {
            return contents == null;
        }

        public void setContents(double amount, boolean isRate, JeiIngredient.Role role) {
            contents = new JeiMekanismJoulesIngredient(
                    amount, (isRate ? "J/t " : "J ") + I18n.format(CbtLang.MEKANISM_LASER), role);
        }

        public JeiUiElement<?> createJeiUiElement(ScreenRegion contRegion) {
            ScreenRegion region = uiPosition.computeRegion(bgTexture.getWidth(), bgTexture.getHeight(), contRegion);
            ScreenRegion barRegion = new ScreenRegion(region.getX() + barOffsetX, region.getY() + barOffsetY,
                    region.getWidth() - barOffsetX * 2, region.getHeight() - barOffsetY * 2);
            return new JeiUiElement<Double>() {
                @Nullable
                @Override
                public JeiIngredient<Double> getIngredient() {
                    return contents;
                }

                @Override
                public ScreenRegion getIngredientRegion() {
                    return barRegion;
                }

                @Override
                public void render(@Nullable Double ingredient) {
                    bgTexture.draw(region.getX(), region.getY());
                    if (ingredient != null) {
                        barOrientation.draw(fgTexture, barRegion.getX(), barRegion.getY(),
                                Double.valueOf(ingredient / capacity).floatValue());
                    }
                }
            };
        }

    }

    public static class JeiAccumulator {

        private final Deque<JeiBuffer> buffers = new LinkedList<>();

        public void accumulate(JeiBuffer buffer) {
            if (buffer.isEmpty()) {
                buffers.offer(buffer);
            }
        }

        public boolean addIngredient(double amount, boolean isRate, JeiIngredient.Role role) {
            while (true) {
                JeiBuffer buffer = buffers.poll();
                if (buffer == null) {
                    return false;
                }
                if (buffer.isEmpty()) {
                    buffer.setContents(amount, isRate, role);
                    return true;
                }
            }
        }

    }

    public static final BufferType<MekanismLaserBuffer, Accumulator, JeiBuffer, JeiAccumulator> TYPE
            = new BufferType<MekanismLaserBuffer, Accumulator, JeiBuffer, JeiAccumulator>() {
        @Override
        public ResourceLocation getId() {
            return ID;
        }

        @Override
        public Class<MekanismLaserBuffer> getBufferClass() {
            return MekanismLaserBuffer.class;
        }

        @Override
        public Class<Accumulator> getAccumulatorClass() {
            return Accumulator.class;
        }

        @Override
        public BufferFactory<MekanismLaserBuffer, JeiBuffer> loadBufferFactory(JsonObject config) {
            double capacity = config.get("capacity").getAsDouble();
            if (capacity <= 0) {
                throw new ConfigException(String.format("Expected positive capacity but got %f", capacity));
            }
            Positioning uiPosition = config.has("position")
                    ? Positioning.fromJson(config.get("position")) : Positioning.FromCenter.CENTER;
            TextureRegion bgTexture = config.has("bar_bg")
                    ? DataLoadUtils.loadTextureRegion(config.get("bar_bg").getAsJsonObject())
                    : CbtTextureResources.MEKANISM_LASER_BAR_BG;
            TextureRegion fgTexture = config.has("bar_fg")
                    ? DataLoadUtils.loadTextureRegion(config.get("bar_fg").getAsJsonObject())
                    : CbtTextureResources.MEKANISM_LASER_BAR_FG;
            int barOffsetX = config.has("bar_offset_x") ? config.get("bar_offset_x").getAsInt() : 1;
            int barOffsetY = config.has("bar_offset_y") ? config.get("bar_offset_y").getAsInt() : 1;
            DrawOrientation barOrientation = config.has("bar_orientation")
                    ? DataLoadUtils.loadDrawOrientation(config.get("bar_orientation").getAsString())
                    : DrawOrientation.BOTTOM_TO_TOP;
            return new BufferFactory<MekanismLaserBuffer, JeiBuffer>() {
                @Override
                public MekanismLaserBuffer createBuffer(World world, BlockPos pos, @Nullable BufferObserver observer) {
                    return new MekanismLaserBuffer(capacity,
                            uiPosition, bgTexture, fgTexture, barOffsetX, barOffsetY, barOrientation, observer);
                }

                @Override
                public JeiBuffer createJeiBuffer() {
                    return new JeiBuffer(capacity,
                            uiPosition, bgTexture, fgTexture, barOffsetX, barOffsetY, barOrientation);
                }
            };
        }

        @Override
        public void configureDefaultSides(MachineSideHandler.SideConfig<MekanismLaserBuffer> sideConfig) {
            for (BlockSide side : BlockSide.VALUES) {
                sideConfig.setEnabled(side, true);
            }
        }

        @Override
        public void attachCapabilities(ICapabilityInstanceConsumer target, MekanismLaserBuffer buffer) {
            target.accept(Capabilities.LASER_RECEPTOR_CAPABILITY, buffer);
        }

        @Override
        public Accumulator createAccumulator() {
            return new Accumulator();
        }

        @Override
        public void accumulate(Accumulator acc, MekanismLaserBuffer buffer) {
            acc.accumulate(buffer);
        }

        @Override
        public Accumulator copyAccumulator(Accumulator acc) {
            return acc.copy();
        }

        @Override
        public UiElement createUiElement(MekanismLaserBuffer buffer) {
            return (gui, index, contRegion) -> {
                TextureRegion bgTex = buffer.getBackgroundTexture();
                ScreenRegion region = buffer.getUiPosition()
                        .computeRegion(bgTex.getWidth(), bgTex.getHeight(), contRegion);
                gui.addComponent(new DoubleEnergyBarGuiComponent(region.getX(), region.getY(),
                        buffer.getBackgroundTexture(), buffer.getForegroundTexture(),
                        buffer.getBarOffsetX(), buffer.getBarOffsetY(), buffer.getBarOrientation(),
                        buffer::getEnergyStored, buffer.getCapacity(), "J"));
                return region;
            };
        }

        @Override
        public void serializeBufferNbt(NBTTagCompound tag, MekanismLaserBuffer buffer) {
            buffer.serNBT(tag);
        }

        @Override
        public void serializeBufferBytes(ByteUtils.Writer stream, MekanismLaserBuffer buffer) {
            buffer.serBytes(stream);
        }

        @Override
        public void deserializeBufferNbt(NBTTagCompound tag, MekanismLaserBuffer buffer) {
            buffer.deserNBT(tag);
        }

        @Override
        public void deserializeBufferBytes(ByteUtils.Reader stream, MekanismLaserBuffer buffer) {
            buffer.deserBytes(stream);
        }

        @Override
        public JeiAccumulator createJeiAccumulator() {
            return new JeiAccumulator();
        }

        @Override
        public void jeiAccumulate(JeiAccumulator acc, JeiBuffer buffer) {
            acc.accumulate(buffer);
        }

        @Override
        public Collection<JeiUiElement<?>> createJeiUiElements(JeiBuffer buffer, ScreenRegion contRegion,
                                                               IJeiHelpers jeiHelpers) {
            return Collections.singletonList(buffer.createJeiUiElement(contRegion));
        }
    };

    @SuppressWarnings("rawtypes")
    @SubscribeEvent
    public static void onRegisterBufferTypes(CbtRegistrationEvent<BufferType> event) {
        event.register(TYPE);
    }

    public static class EnergyMatcher implements IngredientMatcher<Accumulator, JeiAccumulator> {

        public static final ResourceLocation ID = CbtMod.INSTANCE.newResourceLocation("energy");

        private final double amount;
        private final boolean doConsume;

        public EnergyMatcher(double amount, boolean doConsume) {
            this.amount = amount;
            this.doConsume = doConsume;
        }

        @Override
        public boolean consumeInitial(Supplier<Accumulator> acc, float consumeFactor) {
            if (!doConsume) {
                return true;
            }
            double newAmount = consumeFactor * amount;
            if (newAmount <= 0) {
                return true;
            }
            return acc.get().extract(newAmount, true) >= newAmount;
        }

        @Override
        public boolean populateJei(JeiAccumulator acc) {
            return acc.addIngredient(amount, false, JeiIngredient.Role.INPUT);
        }

        @Override
        public Collection<JeiIngredient<?>> getJeiIngredients() {
            return Collections.singletonList(new JeiMekanismJoulesIngredient(
                    amount, "J " + I18n.format(CbtLang.MEKANISM_LASER), JeiIngredient.Role.INPUT));
        }

        public static final IngredientMatcherType<Accumulator, JeiAccumulator> TYPE
                = new IngredientMatcherType<Accumulator, JeiAccumulator>() {
            @Override
            public ResourceLocation getId() {
                return ID;
            }

            @Override
            public IngredientMatcher<Accumulator, JeiAccumulator> loadMatcher(JsonObject config) {
                return new EnergyMatcher(config.get("amount").getAsDouble(),
                        !config.has("consume") || config.get("consume").getAsBoolean());
            }
        };

    }

    public static class PowerMatcher implements IngredientMatcher<Accumulator, JeiAccumulator> {

        public static final ResourceLocation ID = CbtMod.INSTANCE.newResourceLocation("power");

        private final double rate;

        public PowerMatcher(double rate) {
            this.rate = rate;
        }

        @Override
        public boolean consumeInitial(Supplier<Accumulator> acc, float consumeFactor) {
            double newAmount = consumeFactor * rate;
            return newAmount <= 0 || acc.get().getStored() >= newAmount;
        }

        @Override
        public boolean consumePeriodic(Supplier<Accumulator> acc, float consumeFactor) {
            double newAmount = consumeFactor * rate;
            if (newAmount <= 0) {
                return true;
            }
            return acc.get().extract(newAmount, true) >= newAmount;
        }

        @Override
        public boolean populateJei(JeiAccumulator acc) {
            return acc.addIngredient(rate, true, JeiIngredient.Role.INPUT);
        }

        @Override
        public Collection<JeiIngredient<?>> getJeiIngredients() {
            return Collections.singletonList(new JeiMekanismJoulesIngredient(
                    rate, "J/t " + I18n.format(CbtLang.MEKANISM_LASER), JeiIngredient.Role.INPUT));
        }

        public static final IngredientMatcherType<Accumulator, JeiAccumulator> TYPE
                = new IngredientMatcherType<Accumulator, JeiAccumulator>() {
            @Override
            public ResourceLocation getId() {
                return ID;
            }

            @Override
            public IngredientMatcher<Accumulator, JeiAccumulator> loadMatcher(JsonObject config) {
                return new PowerMatcher(config.get("rate").getAsInt());
            }
        };

    }

    @SubscribeEvent
    public static void onRegisterIngredients(CbtIngredientHandlerRegistrationEvent<Accumulator, JeiAccumulator> event) {
        if (event.getBufferType() == TYPE) {
            event.registerMatcherType(EnergyMatcher.TYPE);
            event.registerMatcherType(PowerMatcher.TYPE);
        }
    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/compat/mekanism/buffer/MekanismLaserBuffer.kt
package st.evening.mc.cbtweaker.compat.mekanism.buffer

import mekanism.api.lasers.ILaserReceptor
import mekanism.common.capabilities.Capabilities
import mezz.jei.api.IGuiHelper
import mezz.jei.api.IJeiHelpers
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.EnumFacing
import net.minecraft.util.ResourceLocation
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import st.evening.mc.cbtweaker.CbTweaker
import st.evening.mc.cbtweaker.buffer.BufferFactory
import st.evening.mc.cbtweaker.buffer.BufferObserver
import st.evening.mc.cbtweaker.buffer.BufferType
import st.evening.mc.cbtweaker.buffer.ingredient.IngredientMatcher
import st.evening.mc.cbtweaker.buffer.ingredient.IngredientMatcherType
import st.evening.mc.cbtweaker.compat.jei.ingredient.JeiIngredient
import st.evening.mc.cbtweaker.compat.jei.ui.JeiUiElement
import st.evening.mc.cbtweaker.compat.mekanism.MekanismCompat
import st.evening.mc.cbtweaker.compat.mekanism.jei.JeiMekanismJoulesIngredient
import st.evening.mc.cbtweaker.gui.CbtGuiData
import st.evening.mc.cbtweaker.gui.element.BarControl
import st.evening.mc.cbtweaker.gui.inventory.SyncedUiElement
import st.evening.mc.cbtweaker.gui.inventory.UiElement
import st.evening.mc.cbtweaker.gui.inventory.UiElementWrapper
import st.evening.mc.cbtweaker.util.gui.DrawableData
import st.evening.mc.cbtweaker.util.gui.SamplableData
import st.evening.mc.cbtweaker.util.gui.UiPosition
import st.evening.mc.prelude.api.data.ser.DoubleSerializer
import st.evening.mc.prelude.api.data.ser.NbtCompoundSerializable
import st.evening.mc.prelude.api.data.ser.SerializationException
import st.evening.mc.prelude.api.data.state.Piecewise
import st.evening.mc.prelude.api.data.state.ValueStateAtom
import st.evening.mc.prelude.api.data.tjson.JsonPath
import st.evening.mc.prelude.api.data.tjson.TJson
import st.evening.mc.prelude.api.data.tjson.expectBool
import st.evening.mc.prelude.api.data.tjson.expectDoubleValue
import st.evening.mc.prelude.api.data.tjson.expectInt
import st.evening.mc.prelude.api.data.tjson.useAny
import st.evening.mc.prelude.api.data.tjson.useDoubleValue
import st.evening.mc.prelude.api.data.tjson.useString
import st.evening.mc.prelude.api.gui.drawable.drawFullSize
import st.evening.mc.prelude.api.gui.drawable.drawFullSizeAsProgress
import st.evening.mc.prelude.api.gui.engine.prefab.StackLayout
import st.evening.mc.prelude.api.resource
import st.evening.mc.prelude.api.util.data.runAction
import st.evening.mc.prelude.api.util.game.CapabilityVisitor
import st.evening.mc.prelude.api.util.game.ClientSide
import st.evening.mc.prelude.api.util.game.RequireMod
import st.evening.mc.prelude.api.util.game.ServerSide
import st.evening.mc.prelude.api.util.math.IntRectangle
import st.evening.mc.prelude.api.util.math.Rect2i
import st.evening.mc.prelude.api.util.render.gui.DrawOrientation

@RequireMod(MekanismCompat.MOD_ID)
class MekanismLaserBuffer(
    val config: Config,
    val world: World,
    val bufPos: BlockPos,
    energyStored: Double,
    private val observer: BufferObserver?
) : ILaserReceptor, NbtCompoundSerializable {
    companion object {
        private const val SER_ENERGY: String = "energy"
    }

    private val energyState: ValueStateAtom<Double> = ValueStateAtom(energyStored, DoubleSerializer)

    var energyStored: Double
        get() = energyState.value
        set(value) {
            if (energyState.update(value)) {
                observer?.onIngredientsChanged()
            }
        }

    fun insert(amount: Double, commit: Boolean): Double {
        if (amount <= 0.0) return 0.0
        val energy = energyStored
        val toTransfer = amount.coerceAtMost(config.capacity - energy)
        if (toTransfer <= 0.0) return 0.0
        if (commit) {
            energyStored = energy + toTransfer
        }
        return toTransfer
    }

    fun extract(amount: Double, commit: Boolean): Double {
        if (amount <= 0.0) return 0.0
        val energy = energyStored
        if (energy <= 0.0) return 0.0
        val toTransfer = energy.coerceAtMost(amount)
        if (commit) {
            energyStored = energy - toTransfer
        }
        return toTransfer
    }

    override fun receiveLaserEnergy(energy: Double, face: EnumFacing) {
        energyStored = (energyStored + energy).coerceAtMost(config.capacity)
    }

    override fun canLasersDig(): Boolean = false

    fun copy(observer: BufferObserver?): MekanismLaserBuffer =
        MekanismLaserBuffer(config, world, bufPos, energyStored, observer)

    override fun writeToNbt(dto: NBTTagCompound) {
        dto.runAction {
            SER_ENERGY double energyState.value
        }
    }

    override fun readFromNbt(dto: NBTTagCompound) {
        energyState.setValueFromSync(dto.getDouble(SER_ENERGY))
    }

    private fun createUiElement(): UiElement = UiElementImpl()

    private inner class UiElementImpl : SyncedUiElement {
        override val syncData: Piecewise
            get() = energyState

        @ClientSide.Strong
        override fun addToGuiScreen(uiIndex: Int, layout: StackLayout, baseSlotIndex: Int, wrapper: UiElementWrapper) {
            config.uiPosition.placeElement(
                uiIndex, layout, wrapper,
                BarControl.DoubleTank(
                    config.barBg.drawable,
                    config.barFg.drawable,
                    { energyStored },
                    config.capacity,
                    "J",
                    config.barOrientation,
                    config.barOffsetX,
                    config.barOffsetY
                )
            )
        }
    }

    class Config(
        val capacity: Double,
        val uiPosition: UiPosition,
        val barBg: DrawableData,
        val barFg: SamplableData,
        val barOffsetX: Int,
        val barOffsetY: Int,
        val barOrientation: DrawOrientation
    )

    class Accumulator {
        private val buffers: MutableList<MekanismLaserBuffer> = mutableListOf()

        fun accumulate(buffer: MekanismLaserBuffer) {
            buffers.add(buffer)
        }

        fun getStored(): Double = buffers.sumOf { it.energyStored }

        fun getCapacity(): Double = buffers.sumOf { it.config.capacity }

        fun insert(amount: Double, commit: Boolean): Double {
            var remaining = amount
            buffers.forEach {
                remaining -= it.insert(remaining, commit)
                if (remaining <= 0) return amount
            }
            return amount - remaining
        }

        fun extract(amount: Double, commit: Boolean): Double {
            var remaining = amount
            buffers.forEach {
                remaining -= it.extract(remaining, commit)
                if (remaining <= 0) return amount
            }
            return amount - remaining
        }

        fun copy(): Accumulator {
            val acc = Accumulator()
            buffers.forEach {
                acc.accumulate(it.copy(null))
            }
            return acc
        }
    }

    class JeiBuffer(private val config: Config) {
        private var contents: JeiIngredient<Double>? = null

        fun isEmpty(): Boolean = contents == null

        fun setContents(ingredient: JeiIngredient<Double>) {
            contents = ingredient
        }

        @ClientSide.Physical
        fun createJeiUiElement(contRegion: IntRectangle, guiHelper: IGuiHelper): JeiUiElement<*> {
            val barBg = config.barBg.drawable
            val barFg = config.barFg.drawable
            val pos = config.uiPosition.computePosition(contRegion, barBg.width, barBg.height)
            val barRegion = Rect2i(pos.x + config.barOffsetX, pos.y + config.barOffsetY, barFg.width, barFg.height)
            val ticker = guiHelper.createTickTimer(32, 32, false)
            return object : JeiUiElement<Double> {
                override val jeiIngredient: JeiIngredient<Double>?
                    get() = contents

                override val ingredientRegion: IntRectangle
                    get() = barRegion

                override fun drawElement(ingredient: Double?, partialTicks: Float) {
                    config.barBg.drawable.drawFullSize(partialTicks, pos.x, pos.y)
                    if (ingredient != null) {
                        config.barFg.drawable.drawFullSizeAsProgress(
                            partialTicks,
                            barRegion.posX,
                            barRegion.posY,
                            config.barOrientation,
                            if (contents?.role == JeiIngredient.Role.OUTPUT) {
                                ticker.value / ticker.maxValue.toFloat()
                            } else {
                                1 - ticker.value / ticker.maxValue.toFloat()
                            }
                        )
                    }
                }
            }
        }
    }

    class JeiAccumulator {
        private val buffers: ArrayDeque<JeiBuffer> = ArrayDeque()

        fun accumulate(buffer: JeiBuffer) {
            if (buffer.isEmpty()) {
                buffers += buffer
            }
        }

        fun addIngredient(ingredient: JeiIngredient<Double>): Boolean {
            while (buffers.isNotEmpty()) {
                val buffer = buffers.removeFirst()
                if (buffer.isEmpty()) {
                    buffer.setContents(ingredient)
                    return true
                }
            }
            return false
        }
    }

    object Type : BufferType<MekanismLaserBuffer, Accumulator, JeiBuffer, JeiAccumulator> {
        override val id: ResourceLocation = CbTweaker.resource("mekanism_laser")

        override val bufferClass: Class<MekanismLaserBuffer>
            get() = MekanismLaserBuffer::class.java
        override val accumulatorClass: Class<Accumulator>
            get() = Accumulator::class.java

        context(_: JsonPath)
        override fun loadBufferFactory(dto: TJson.Object): BufferFactory<MekanismLaserBuffer, JeiBuffer> {
            val capacity = dto.useDoubleValue("capacity") {
                if (it < 0.0) throw SerializationException.withPath("Capacity must be positive!")
                return@useDoubleValue it
            }
            val config = Config(
                capacity,
                dto.useAny("ui_position") { UiPosition.load(it) } ?: UiPosition.CENTER,
                dto.useAny("bar_bg") { DrawableData.loadSliceOrBlank(it, 6, 36) } ?: CbtGuiData.MEKANISM_ENERGY_BAR_BG,
                dto.useAny("bar_fg") { DrawableData.loadSliceOrBlank(it, 4, 34) } ?: CbtGuiData.MEKANISM_ENERGY_BAR_FG,
                dto.expectInt("bar_offset_x") ?: 1,
                dto.expectInt("bar_offset_y") ?: 1,
                dto.useString("bar_orientation") { DrawOrientation.serializer.deserializeFromJson(it) }
                    ?: DrawOrientation.BOTTOM_TO_TOP
            )
            return object : BufferFactory<MekanismLaserBuffer, JeiBuffer> {
                override fun createBuffer(world: World, pos: BlockPos, observer: BufferObserver): MekanismLaserBuffer =
                    MekanismLaserBuffer(config, world, pos, 0.0, observer)

                override fun createJeiBuffer(): JeiBuffer = JeiBuffer(config)
            }
        }

        override fun attachCapabilities(target: CapabilityVisitor, buffer: MekanismLaserBuffer) {
            target.visit(Capabilities.LASER_RECEPTOR_CAPABILITY, buffer)
        }

        override fun createAccumulator(): Accumulator = Accumulator()

        override fun accumulate(acc: Accumulator, buffer: MekanismLaserBuffer) {
            acc.accumulate(buffer)
        }

        override fun copyAccumulator(acc: Accumulator): Accumulator = acc.copy()

        @ServerSide
        override fun serializeBufferToNbt(buffer: MekanismLaserBuffer, dto: NBTTagCompound) {
            buffer.writeToNbt(dto)
        }

        @ServerSide
        override fun deserializeBufferFromNbt(buffer: MekanismLaserBuffer, dto: NBTTagCompound) {
            buffer.readFromNbt(dto)
        }

        override fun createUiElement(buffer: MekanismLaserBuffer): UiElement = buffer.createUiElement()

        override fun createJeiAccumulator(): JeiAccumulator = JeiAccumulator()

        override fun jeiAccumulate(acc: JeiAccumulator, buffer: JeiBuffer) {
            acc.accumulate(buffer)
        }

        @ClientSide.Physical
        override fun createJeiUiElements(
            buffer: JeiBuffer,
            contRegion: IntRectangle,
            jeiHelpers: IJeiHelpers
        ): Collection<JeiUiElement<*>> = listOf(buffer.createJeiUiElement(contRegion, jeiHelpers.guiHelper))
    }

    class EnergyMatcher(private val amount: Double, private val doConsume: Boolean) :
        IngredientMatcher<Accumulator, JeiAccumulator> {

        override fun consumeInitial(acc: Lazy<Accumulator>, consumeFactor: Float, determMode: Boolean): Boolean {
            val scaledAmount = amount * consumeFactor
            return scaledAmount <= 0.0 || acc.value.extract(scaledAmount, !doConsume) >= scaledAmount
        }

        private val jeiIngredient: JeiMekanismJoulesIngredient =
            JeiMekanismJoulesIngredient(amount, false, JeiIngredient.Role.INPUT)

        override fun getJeiIngredients(): Collection<JeiIngredient<*>> = listOf(jeiIngredient)

        override fun populateJei(acc: JeiAccumulator): Boolean = acc.addIngredient(jeiIngredient)

        object Type : IngredientMatcherType<Accumulator, JeiAccumulator> {
            override val id: ResourceLocation = CbTweaker.resource("energy")

            context(_: JsonPath)
            override fun loadMatcher(dto: TJson.Object): EnergyMatcher =
                EnergyMatcher(dto.expectDoubleValue("amount"), dto.expectBool("consume") ?: true)
        }
    }

    class PowerMatcher(private val rate: Double) : IngredientMatcher<Accumulator, JeiAccumulator> {
        override fun consumeInitial(acc: Lazy<Accumulator>, consumeFactor: Float, determMode: Boolean): Boolean =
            consume(acc, consumeFactor, false) // make sure there's enough energy to start

        override fun consumePeriodic(acc: Lazy<Accumulator>, consumeFactor: Float): Boolean =
            consume(acc, consumeFactor, true)

        private fun consume(acc: Lazy<Accumulator>, consumeFactor: Float, commit: Boolean): Boolean {
            val scaledAmount = rate * consumeFactor
            return scaledAmount <= 0 || acc.value.extract(scaledAmount, commit) >= scaledAmount
        }

        private val jeiIngredient: JeiMekanismJoulesIngredient =
            JeiMekanismJoulesIngredient(rate, true, JeiIngredient.Role.INPUT)

        override fun getJeiIngredients(): Collection<JeiIngredient<*>> = listOf(jeiIngredient)

        override fun populateJei(acc: JeiAccumulator): Boolean = acc.addIngredient(jeiIngredient)

        object Type : IngredientMatcherType<Accumulator, JeiAccumulator> {
            override val id: ResourceLocation = CbTweaker.resource("power")

            context(_: JsonPath)
            override fun loadMatcher(dto: TJson.Object): PowerMatcher = PowerMatcher(dto.expectDoubleValue("rate"))
        }
    }
}
